src/ooo_cpu.cc

```diff
@@ -304,11 +304,58 @@ void O3_CPU::fetch_instruction()
       return;
     }
 
-  // scan through IFETCH_BUFFER to find instructions that need to be translated
+  std::size_t available_fetch_bandwidth = FETCH_WIDTH;
+
+  // scan through IFETCH_BUFFER to find instructions that hit in the decoded instruction buffer
   uint32_t index = IFETCH_BUFFER.head;
   for(uint32_t i=0; i<IFETCH_BUFFER.SIZE; i++)
     {
-        ooo_model_instr &ifb_entry = IFETCH_BUFFER.entry[index];
+      ooo_model_instr &ifb_entry = IFETCH_BUFFER.entry[index];
+
+      // Check DIB to see if we recently fetched this line
+      dib_t::value_type &dib_set = DIB[(ifb_entry.ip >> LOG2_BLOCK_SIZE) % DIB_SET];
+      auto way = std::find_if(dib_set.begin(), dib_set.end(), [ifb_entry](dib_entry_t x){ return x.valid && ((x.addr >> LOG2_BLOCK_SIZE) == (ifb_entry.ip >> LOG2_BLOCK_SIZE));});
+      if (way != dib_set.end())
+      {
+          // The cache line is in the L0, so we can mark this as complete
+          ifb_entry.fetched = COMPLETED;
+
+          // Also mark it as decoded
+          ifb_entry.decoded = COMPLETED;
+
+          // It can be acted on immediately
+          ifb_entry.event_cycle = current_core_cycle[cpu];
+
+          // Update LRU
+          unsigned hit_lru = way->lru;
+          std::for_each(dib_set.begin(), dib_set.end(), [hit_lru](dib_entry_t &x){ if (x.lru <= hit_lru) x.lru++; });
+          way->lru = 0;
+
+	  available_fetch_bandwidth--;
+      }
+
+      if(available_fetch_bandwidth == 0)
+	{
+	  break;
+	}
+
+      index++;
```

Comment:
This early advance is potentially pretty dangerous. It means that we will not allow two consecutive entries in the IFETCH_BUFFER to be checked in the DIB. The way this appears to work is 

1. IFETCH_BUFFER[head] is checked in DIB.
    - index is incremented
2. IFETCH_BUFFER[head+1] is checked for translation.
    - IFETCH_BUFFER[head+1] was never checked in DIB.
    - index is incremented on L481
3. IFETCH_BUFFER[head+2] is checked in DIB
... and so on

This increment and check should be removed and we should rely on L481 to increment the index.

Reply:
This is an entirely separate for-loop.  I broke out the DIB-related stuff into its own for-loop that runs before the translate/fetch loop.  This was done to make it clear that this is an alternative path that instructions can go down that is independent of the main translate/fetch path.

Reply:
Now that I opened it in a proper editor, I can see that. It wasn't clear from the git diff. Just to put the thought out there, I think before too long I want to break `O3_CPU::fetch_instruction()` up into three functions: `O3_CPU::check_dib()`, `O3::CPU::translate_instruction()`, and `O3_CPU::fetch_instruction()`. This would increase readability, since it better describes what's going on. That is beyond the scope of this patch, though.